src/sinks/datadog/logs/service.rs

```diff
@@ -144,12 +150,14 @@ impl HttpSink for Service {
             .header("Content-Type", "application/json")
             .header("DD-API-KEY", &api_key[..]);
 
-        let (request, body) = match self.compression {
+        let (request, encoded_body) = match self.compression {
             Compression::None => (request, body),
             Compression::Gzip(level) => {
                 let level = level.unwrap_or(GZIP_FAST);
-                let mut encoder =
-                    GzEncoder::new(Vec::new(), flate2::Compression::new(level as u32));
+                let mut encoder = GzEncoder::new(
+                    Vec::with_capacity(body.len()),
+                    flate2::Compression::new(level as u32),
+                );
```

Comment:
This makes me wish that there was a way for compressors to compress _into_ the existing allocation, instead of only using a small stack allocated buffer for themselves to shuffle blocks around.  Since compressed values should always be smaller than the input, you could imagine reading a 64kb block or whatever, compressing it, and writing back over that block in the source vector with the compressed bit.  Like the inverse of a circular buffer.

Maybe it doesn't actually matter in practice, just a random thought. :P

Reply:
I feel very much the same. It doesn't matter much in practice for us yet but I can easily imagine a situation where it will. 

What you describe is interesting and not something I'd considered. I like it.

Reply:
Actually, compressed values may in fact be larger than the original. Consider the case of uncompressible data. In that case, the compressor has to add a header with at least a flag that the data is not compressed, resulting in an output larger than the input.